src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -162,6 +508,105 @@ CQueryMutators::NormalizeGroupByProjList(CMemoryPool *mp,
 {
 	Query *query_copy = (Query *) gpdb::CopyObject(const_cast<Query *>(query));
 
+	// If there is a functional dependency proved for a relation on a set of
+	// grouping columns, then it is valid to have a column in the target list
+	// even if it is not listed explicitly in the groupClause. To make ORCA
+	// correctly process such cases, we add all such target list entries into
+	// groupClause's grouping sets, where the functional dependency exists.
+	// If there is such functional dependency existing, the OID of
+	// Primary Key constraint resides in the constraintDeps list. As the
+	// constraintDeps list contains only Primary Key OIDs, we skip checks of
+	// constraint type in all code below.
+	// Refer to function check_functional_grouping for more details.
+	if (0 != gpdb::ListLength(query_copy->constraintDeps))
+	{
+		// Explicit addition of functionally dependent columns requires the
+		// following steps:
+		// 1. Store current unique TargetlistEntry references in groupClause -
+		// they will be needed at step 3.
+		// 2. Update all grouping sets that contain Primary Key
+		// 3. Update arguments of GROUPING functions, because they could be
+		// shifted after step 3.
+
+		// Step 1.
+		// Store current unique TargetlistEntry references in groupClause.
+		// After modifying groupClause, they will be required to update
+		// arguments inside grouping functions.
+		SContexGetGroupUniqueTleWalker ctx_old_grouping_tle_refs;
+		GetGroupUniqueTargetlistEntriesWalker((Node *) query_copy->groupClause,
+											  &ctx_old_grouping_tle_refs);
+
+		// Step 2.
+		// Update groupClause - add all functionally dependent entries explicitly.
+		ListCell *lc_constraint = NULL;
+		ForEach(lc_constraint, query_copy->constraintDeps)
+		{
+			Oid constraint_oid = lfirst_oid(lc_constraint);
+			Oid conrelid = 0;
+			Oid confrelid =
+				0;	// not used, but function get_constraint_relation_oids requires it..
+			gpdb::GetConstraintRelationOids(constraint_oid, &conrelid,
+											&confrelid);
+			List *conkeys_list =
+				gpdb::GetConstraintRelationColumns(constraint_oid);
+			SContextAddMissingGroupClause ctx_fix_group_clause(
+				query_copy, conrelid, conkeys_list);
+
+			AddMissingGroupClauseWalker((Node *) query_copy->targetList,
```

Comment:
And, what happens when we add the SubLink into the TL?

```
SELECT b, (select c from (select c) q1) FROM test_table1 GROUP BY a ORDER BY a;
INFO:  GPORCA failed to produce a plan, falling back to planner
DETAIL:  Query-to-DXL Translation: No attribute entry found due to incorrect normalization of query


```

Reply:
> Maybe we return to previous variant (it seems more intuitive)

Actualy, I even had to do it, to fix issue with SubLink below.

> Step 1: Use walker (which has been removed, or import gpdb::flatten_tlist and then filter out extras) to extract the Vars that are not explicitly in TL, and put them to a separate list (maybe add that list to SContextAddMissingGroupClause) without concatenating that list with original TL.

We can't just call flatten_tlist, as it doesn't expect to have varlevelsup != 0 (it expects that input tlist is already preprocessed). And pulling all preprocessing code from ```grouping_planner``` doesn't sound inspiring :)
Storing resjunk entries in separate list, then adding them back to query's TL seems a bit more complicated then immediately add to the query's TL.
Thus, I think better to get back the original walker.


> And i also think we could skip the check  CTranslatorUtils::IsGroupingColumn, because the grouping columns are already in original TL, and new entries are not in the original TL.

We can't, because there could be explicit and non-explicit grouping columns in the original TL. For ex., query:
```
SELECT a, b, c FROM test_table1 GROUP BY a, b ORDER BY a;
```
We should add a new SortGroupClause for 'c', but shouldn't do it for 'b' (we do not want duplicates). CTranslatorUtils::IsGroupingColumn will filter out 'b'.

Reply:
>Storing resjunk entries in separate list, then adding them back to query's TL seems a bit more complicated then immediately add to the query's TL.

Firstly, give me, please, an example of a walker, which modifies the original structure the walker traverses. 

Secondly, I really don't really understand the AddMissingGroupClauseWalker. We don't need a recursion. We don't need a walker to just process flat TagetEntries with vars. We can simply iterate that in obvious manner. Why do we need to walk the target list?

Reply:
let me double check it

Reply:
>Storing resjunk entries in separate list, then adding them back to query's TL seems a bit more complicated then immediately add to the query's TL.

And about that. Aren't things like below looking weird?
```
foreach(lc, targetList)
{
  if (condidtion) targetList = lappend(targetList, newEntry)
}
```
I mean we are modifying the list we are traversing inside the ExtractVarsIntoTargetlistWalker. I don't know, maybe its normal. But this is definetely not obvious, walkers are not mutators, they are not modifying anything.

Reply:
I've just re-checked expression_tree_walker.
According to its documentation:
```
 * expression_tree_walker() is designed to support routines that traverse
 * a tree in a read-only fashion (although it will also work for routines
 * that modify nodes in-place but never add/delete/replace nodes)
```
Thought current approach with adding target entries in-place is working, it contradicts the philosophy of expression_tree_walker.
I will evaluate the possibility to avoid in-place adding of target entries.

Reply:
Ok, I've removed AddMissingGroupClauseWalker. Indeed now we can leave with simple loop over direct entries of target list, without recursing into it. 
Also revised the walker that extracted vars without TLEs from expressions. Now it stores new TLEs in a list, that is later concatenated to target list. Thus in-place insertion is avoided inside the walker.